torch/csrc/cuda/shared/cudnn.cpp

```diff
@@ -0,0 +1,94 @@
+// The clang-tidy job seems to complain that it can't find cudnn.h without this.
+// This file should only be compiled if this condition holds, so it should be safe.
+#if defined(USE_CUDNN) || defined(__HIP_PLATFORM_HCC__)
+#include <torch/csrc/utils/pybind.h>
+
+#include <array>
+#include <tuple>
+
+namespace {
+using version_tuple = std::tuple<size_t, size_t, size_t>;
+}
+
+#ifdef USE_CUDNN
+#include <cudnn.h>
+
+namespace {
+
+version_tuple getCompileVersion() {
+  return version_tuple(CUDNN_MAJOR, CUDNN_MINOR, CUDNN_PATCHLEVEL);
+}
+
+version_tuple getRuntimeVersion() {
+  auto version = cudnnGetVersion();
+  auto major = version / 1000;
+  auto minor = (version % 1000) / 100;
+  auto patch = version % 10;
+  return version_tuple(major, minor, patch);
+}
+
+size_t getVersionInt() {
+  return cudnnGetVersion();
+}
+
+}
+#elif defined(__HIP_PLATFORM_HCC__)
+#include <miopen/miopen.h>
+#include <miopen/version.h>
+
+namespace {
+
+version_tuple getCompileVersion() {
+  return version_tuple(
+      MIOPEN_VERSION_MAJOR, MIOPEN_VERSION_MINOR, MIOPEN_VERSION_PATCH);
+}
+
+version_tuple getRuntimeVersion() {
+  // MIOpen doesn't include runtime version info before 2.3.0
+#if (MIOPEN_VERSION_MAJOR > 2) || (MIOPEN_VERSION_MAJOR == 2 && MIOPEN_VERSION_MINOR > 2)
+  size_t major, minor, patch;
+  miopenGetVersion(&major, &minor, &patch);
+  return version_tuple(major, minor, patch);
+#else
+  return getCompileVersion();
+#endif
+}
+
+size_t getVersionInt() {
+  // miopen version is MAJOR*1000000 + MINOR*1000 + PATCH
+  size_t major, minor, patch;
+  std::tie(major, minor, patch) = getRuntimeVersion();
+  return major * 1000000 + minor * 1000 + patch;
+}
+
+}
+#endif
+
+namespace torch { namespace cuda { namespace shared {
+
+void initCudnnBindings(PyObject* module) {
+  auto m = py::handle(module).cast<py::module>();
+
+  auto cudnn = m.def_submodule("_cudnn", "libcudnn.so bindings");
+
+  py::enum_<cudnnRNNMode_t>(cudnn, "RNNMode")
+    .value("rnn_relu", CUDNN_RNN_RELU)
+    .value("rnn_tanh", CUDNN_RNN_TANH)
+    .value("lstm", CUDNN_LSTM)
+    .value("gru", CUDNN_GRU);
+
+#ifdef USE_CUDNN
+  cudnn.attr("is_cuda") = true;
```

Comment:
This... doesn't make any sense at all! XD Did this come from the original patch?

Reply:
>Did this come from the original patch?

No, none of the miopen stuff was in the original patch. It's maybe a bad name but I need to distinguish between cudnn and miopen for the version check.

Reply:
worth a comment!